### PR TITLE
NEEM Hub Configuration

### PR DESCRIPTION
**Pull request by AbhijitVyas**
This PR should be considered with:

- https://github.com/ease-crc/openease/pull/15
- https://github.com/ease-crc/openease_dockerbridge/pull/5

### CODE REVIEW

File: webrob/AlchemyEncoder.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from sqlalchemy.ext.declarative import DeclarativeMeta
+import json
+
+class AlchemyEncoder(json.JSONEncoder):
```
Comment (daniel86) [inline webrob/AlchemyEncoder.py:5]: could you add a comment what the purpose of this class is?

File: webrob/app_and_db.py [hunk 2]
```diff
@@ -7,29 +7,52 @@
...
+
+            mongoDBMetaCollection = mongoDbClient["meta"]
+
+            if mongoDBMetaCollection.count() > 0:
+                app.logger.info('------------ mongoDb collection contains some document values ------------')
+                return mongoDBMetaCollection
+            else:
+                app.logger.info('---  MongoDB connection is established but collection does not contain any values  ---')
+                return mongoDBMetaCollection
+        except:
```
Comment (daniel86) [inline webrob/app_and_db.py]: could you print the exception? else no one will be able to figure out why the connection has failed
I am sure app.logger has an interface for logging exceptions

File: webrob/models/NEEMHubSettings.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from webrob.app_and_db import db, app
+from sqlalchemy.exc import SQLAlchemyError
+
+class NEEMHubSettings(db.Model):
```
Comment (daniel86) [inline webrob/models/NEEMHubSettings.py:10]: comment is missing

File: webrob/models/NEEMHubSettings.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    id = db.Column(db.Integer, primary_key=True)
+    MONGO_HOST = db.Column(db.String(255), nullable=False, default='')
+    MONGO_PORT = db.Column(db.Integer())
+    MONGO_DB = db.Column(db.String(255), nullable=False, default='')
+    MONGO_USER = db.Column(db.String(255), nullable=False, default='')
+    MONGO_PASS = db.Column(db.String(255), nullable=False, default='')
+
+
+def get_settings(id):
```
Comment (daniel86) [inline webrob/models/NEEMHubSettings.py]: method should have docu, e.g. what is the id parameter?

File: webrob/models/NEEMHubSettings.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    try:
+        values = NEEMHubSettings.query.filter_by(id=id).one()
+        return values
+
+    except SQLAlchemyError as e:
+        app.logger.info("get_settings: while connecting to sql db returns null")
+        return NEEMHubSettings()
+
+def get_settings_count():
```
Comment (daniel86) [inline webrob/models/NEEMHubSettings.py:30]: what is the purpose of this? I do not get why there could be multiple settings

File: webrob/neems/manager.py [hunk 6]
```diff
@@ -16,8 +17,12 @@ def __init__(self):
     def __get_neem_ids__(self):
         neem_ids = []
         # get all neem ids information from collection
-        neem_ids = mongoDBMetaCollection.find().distinct('_id')
-        return neem_ids
+        neemHubSettings = get_settings(1)
```
Comment (daniel86) [inline webrob/neems/manager.py]: strange magic number 1

File: webrob/app_and_db.py [hunk 7]
```diff
@@ -7,29 +7,52 @@
...
+os.environ['POSTGRES_PORT_5432_TCP_ADDR'] + ':' + \
+os.environ['POSTGRES_PORT_5432_TCP_PORT'] + '/docker'
 
 # This is the SQLAlchemy ORM object
 db = SQLAlchemy(app)
+
+
+# This method will check db connection with given settings.
+# If settings are correct then will return meta collection from the db
+def checkConnection(neemHubSettings):
```
Comment (daniel86) [inline webrob/app_and_db.py]: the naming is confusing, methods that return some object should have *get* prefix.
prefix *check* suggests to me that the function returns a boolean indicating whether the check succeeds or not.

File: webrob/neems/manager.py [hunk 8]
```diff
@@ -41,5 +46,7 @@ def filter_neems(self, query_string):
         #   - can an array be added to text index?
         #   - regex search possible, but only without index!!
         #       db.meta.find({"keywords": {"$regex": ".*robot.*","$options": 'i'}},{_id: 1}).pretty()
-        return map(lambda doc: doc["_id"], mongoDBMetaCollection.find(
-            {"$text": {"$search": query_string}}, {"_id": 1}))
+        neemHubSettings = get_settings(1)
+        if checkConnection(neemHubSettings) is not None:
+            return map(lambda doc: doc["_id"], checkConnection(neemHubSettings).find(
```
Comment (daniel86) [inline webrob/neems/manager.py]: this code is not optimal. each call of *checkConnection* establishes connection to mongo, so you would rather write something like that:

```
mongo = getMongo(neemHubSettings)
if mongo is not None:
    do_something_with(mongo)
```

File: webrob/pages/QA.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
 
 @app.route('/QA')
 def render_QA_page():
+    neem_manager = NEEM_Manager()
```
Comment (daniel86) [inline webrob/pages/QA.py]: this looks like an error. NEEM Manager is supposed to be a singleton with only one global instance.

File: webrob/pages/QA.py [hunk 10]
```diff
@@ -31,6 +32,7 @@ def render_QA_page():
 
 @app.route('/video')
 def render_video_page():
+    neem_manager = NEEM_Manager()
```
Comment (daniel86) [inline webrob/pages/QA.py]: this looks like an error. NEEM Manager is supposed to be a singleton with only one global instance.

File: webrob/pages/admin/db.py [hunk 11]
```diff
@@ -39,8 +46,8 @@ def db_page_user_roles():
 def db_find_user_roles():
     users = _get_all_users()
     user_roles = []
-    for u in users:     # find user roles
-        user_roles.append({
+    for u in users:  # find user roles
+        user_roles.neemsappend({
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: what is neemsappend?

File: webrob/pages/admin/db.py [hunk 12]
```diff
@@ -145,3 +152,39 @@ def db_remove_route(table):
     if data is not None and 'id' in data and db_find(cls, data['id']):
         db_remove(cls, data['id'])
     return jsonify(result=None)
+
+
+@app.route('/db/page/get_neem_hub_settings')
+@admin_required
+def render_neem_hub_settings_page_get():
+    form = PasswordForm()
```
Comment (daniel86) [inline webrob/pages/admin/db.py:167]: why PasswordForm? why is this needed? comment should be added to explain

File: webrob/pages/neems.py [hunk 13]
```diff
@@ -1,16 +1,28 @@
...
 
 from webrob.neems.manager import NEEM_Manager
 
+from webrob.models.NEEMHubSettings import get_settings_count, get_settings
+
+
 __author__ = 'danielb@cs.uni-bremen.de'
 
-neem_manager = NEEM_Manager()
+
```
Comment (daniel86) [inline webrob/pages/neems.py]: why?

**Review by daniel86 — CHANGES_REQUESTED** [review]

File: webrob/models/NEEMHubSettings.py (hunk 5 above)
Comment (AbhijitVyas) [reply webrob/models/NEEMHubSettings.py:30]: removed

File: webrob/neems/manager.py (hunk 6 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py]: updated.

File: webrob/app_and_db.py (hunk 7 above)
Comment (AbhijitVyas) [reply webrob/app_and_db.py]: renamed method to getMongoDBMetaCollection

File: webrob/models/NEEMHubSettings.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    MONGO_PASS = db.Column(db.String(255), nullable=False, default='')
+
+# get setting values from db(with id==1)
+def get_settings():
+    try:
+        values = NEEMHubSettings.query.filter_by(id=settings_Id).one()
+        return values
+
+    except SQLAlchemyError as e:
+        app.logger.info("get_settings: while connecting to sql db returns null")
```
Comment (daniel86) [inline webrob/models/NEEMHubSettings.py]: the exception is not printed

File: webrob/app_and_db.py [hunk 15]
```diff
@@ -7,29 +7,59 @@
...
+
+            mongoDBMetaCollection = mongoDbClient["meta"]
+
+            if mongoDBMetaCollection.count() > 0:
+                app.logger.info('------------ mongoDb collection contains some document values ------------')
+                return mongoDBMetaCollection
+            else:
+                app.logger.info('---  MongoDB connection is established but collection does not contain any values  ---')
+                return mongoDBMetaCollection
+        except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/app_and_db.py]: this does not look like proper exception handling. The caller has to handle the case that None is returned. So why don't we raise the exception to the caller and et the caller handle it?

File: webrob/app_and_db.py [hunk 16]
```diff
@@ -7,29 +7,59 @@
...
+            app.logger.error('------------ mongoDb connection can not be created ------------')
+            app.logger.error(e)
+            return None
+        except PyMongoError as e:
```
Comment (daniel86) [inline webrob/app_and_db.py]: this does not look like proper exception handling. The caller has to handle the case that None is returned. So why don't we raise the exception to the caller and et the caller handle it?

File: webrob/neems/manager.py [hunk 17]
```diff
@@ -1,23 +1,29 @@
...
-        neem_ids = []
+    def __set_neem_ids__(self):
+        self.neem_ids = []
         # get all neem ids information from collection
-        neem_ids = mongoDBMetaCollection.find().distinct('_id')
-        return neem_ids
+        neemHubSettings = get_settings()
+
+        mongoDBMetaCollection = getMongoDBMetaCollection(neemHubSettings)
+        if mongoDBMetaCollection is not None:
```
Comment (daniel86) [inline webrob/neems/manager.py]: better make a try .. catch ... instead of checking for none.

File: webrob/neems/manager.py [hunk 18]
```diff
@@ -1,23 +1,29 @@
...
+        self.neem_ids = self.__set_neem_ids__()
 
-    def __get_neem_ids__(self):
-        neem_ids = []
+    def __set_neem_ids__(self):
+        self.neem_ids = []
         # get all neem ids information from collection
-        neem_ids = mongoDBMetaCollection.find().distinct('_id')
-        return neem_ids
+        neemHubSettings = get_settings()
```
Comment (daniel86) [inline webrob/neems/manager.py]: neemHubSettings could be None, better check for exceptions instead

File: webrob/neems/neem.py [hunk 19]
```diff
@@ -18,6 +20,8 @@ def __init__(self,
             b_id = bson.objectid.ObjectId(neem_id)
         else:
             b_id = neem_id
+        neemHubSettings = get_settings()
+        mongoDBMetaCollection = getMongoDBMetaCollection(neemHubSettings)
```
Comment (daniel86) [inline webrob/neems/neem.py]: it seems that at every place you call getMongoDBMetaCollection, get_settings is called before. So I would suggest to do this withing the getMongoDBMetaCollection call

File: webrob/app_and_db.py [hunk 20]
```diff
@@ -7,29 +7,59 @@
...
+os.environ['POSTGRES_PORT_5432_TCP_ADDR'] + ':' + \
+os.environ['POSTGRES_PORT_5432_TCP_PORT'] + '/docker'
 
 # This is the SQLAlchemy ORM object
 db = SQLAlchemy(app)
+
+
+# This method will check db connection with given settings.
+# If settings are correct then will return meta collection from the db
+def getMongoDBMetaCollection(neemHubSettings):
```
Comment (daniel86) [inline webrob/app_and_db.py]: shouldn't it be *get_mongo_db_meta_collection* ?
We should be consistent

File: webrob/pages/admin/db.py [hunk 21]
```diff
@@ -145,3 +154,50 @@ def db_remove_route(table):
     if data is not None and 'id' in data and db_find(cls, data['id']):
         db_remove(cls, data['id'])
     return jsonify(result=None)
+
+
+@app.route('/db/page/get_neem_hub_settings')
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: the name of the route is confusing. Better call it *neem_hub_settings* or *edit_neem_hub_settings* or so

File: webrob/pages/admin/db.py [hunk 22]
```diff
@@ -145,3 +154,50 @@ def db_remove_route(table):
...
+        if mongoDBMetaCollection is None:
+            app.logger.error('------------ mongoDb connection can not be created ------------')
+            flash('Failure connecting with mongodb with given credentials, please check inputs!', "warning")
+            return redirect(url_for('render_neem_hub_settings_page_get'))
+        else:
+            # if connection is secured then update neem_ids from mongodb meta collection so that neem discovery page has latest updates
+            neem_manager.__set_neem_ids__()
+            app.logger.info('------------ neem manager updated ------------')
+    else:
+        flash('Form submission causes problems!', "warning")
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: this is not a very helpful message for the user. What is the real problem?

File: webrob/pages/admin/db.py [hunk 23]
```diff
@@ -145,3 +154,50 @@ def db_remove_route(table):
...
+        app.logger.info('Configuration has been stored!')
+        mongoDBMetaCollection = getMongoDBMetaCollection(neemHubSettings)
+
+        if mongoDBMetaCollection is None:
+            app.logger.error('------------ mongoDb connection can not be created ------------')
+            flash('Failure connecting with mongodb with given credentials, please check inputs!', "warning")
+            return redirect(url_for('render_neem_hub_settings_page_get'))
+        else:
+            # if connection is secured then update neem_ids from mongodb meta collection so that neem discovery page has latest updates
+            neem_manager.__set_neem_ids__()
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: why starts this method with two underscore?

**Review by daniel86 — CHANGES_REQUESTED** [review]

File: webrob/pages/admin/db.py (hunk 23 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: no apparent reason. the previous method __get_neem_ids__ was renamed as __set_neem_ids__
I have removed the '__' from method name.

File: webrob/neems/manager.py [hunk 24]
```diff
@@ -1,23 +1,39 @@
...
+    def set_neem_ids(self):
+        self.neem_ids = []
 
-    def __get_neem_ids__(self):
-        neem_ids = []
         # get all neem ids information from collection
-        neem_ids = mongoDBMetaCollection.find().distinct('_id')
-        return neem_ids
+        # first find if there is any existing entry in db
+        count = get_settings_count()
```
Comment (daniel86) [inline webrob/neems/manager.py:19]: why do we test the count here? we would get an exception in get_mongo_db_meta_collection anyways or? so the check is not needed

File: webrob/pages/admin/db.py [hunk 25]
```diff
@@ -145,3 +160,80 @@ def db_remove_route(table):
...
+        db.session.commit()
+        app.logger.info('Configuration has been stored!')
+
+        try:
+            get_mongo_db_meta_collection()
+            # if connection is secured then update neem_ids from mongodb meta collection so that neem discovery page has latest updates
+            neem_manager.set_neem_ids()
+            app.logger.debug('------------ neem manager updated ------------')
+
+        except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: why not adding a flask handler for this error?

File: webrob/pages/admin/db.py [hunk 26]
```diff
@@ -145,3 +160,80 @@ def db_remove_route(table):
...
+            raise NEEMMetaException(
+                'An exception has occurred during connection with mongodb collection, please check!', exc=e)
+
+        except PyMongoError as e:
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: why not adding a flask handler for this error?

File: webrob/pages/admin/db.py [hunk 27]
```diff
@@ -145,3 +160,80 @@ def db_remove_route(table):
...
+@app.route('/edit_neem_hub_settings')
+@admin_required
+def render_neem_hub_settings_page_get():
+    # PasswordForm used for validating given password field
+    form = PasswordForm()
+    try:
+        neemHubSettings = get_settings()
+    except SQLAlchemyError as e:
+        raise SQLAlchemyErrorException('while connecting to sql db raises an exception', exc=e)
+    except NoResultFound as e:
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: why not adding a flask handler for this error?

File: webrob/pages/admin/db.py [hunk 28]
```diff
@@ -145,3 +160,80 @@ def db_remove_route(table):
...
+
+
+@app.route('/edit_neem_hub_settings')
+@admin_required
+def render_neem_hub_settings_page_get():
+    # PasswordForm used for validating given password field
+    form = PasswordForm()
+    try:
+        neemHubSettings = get_settings()
+    except SQLAlchemyError as e:
```
Comment (daniel86) [inline webrob/pages/admin/db.py]: why not adding a flask handler for this error?

File: webrob/neems/neem.py [hunk 29]
```diff
@@ -61,9 +76,17 @@ def get_info(self):
 
     def activate(self):
         app.logger.info('Activate neem')
-        if not USE_HOST_KNOWROB and not container_started(current_user.username):
+        try:
+            neemHubSettings = get_settings()
+        except SQLAlchemyError as e:
+            raise SQLAlchemyErrorException('while connecting to sql db raises an exception', exc=e)
+        except NoResultFound as e:
```
Comment (daniel86) [inline webrob/neems/neem.py]: why not adding a flask handler for this error?

File: webrob/neems/neem.py [hunk 30]
```diff
@@ -61,9 +76,17 @@ def get_info(self):
 
     def activate(self):
         app.logger.info('Activate neem')
-        if not USE_HOST_KNOWROB and not container_started(current_user.username):
+        try:
+            neemHubSettings = get_settings()
+        except SQLAlchemyError as e:
```
Comment (daniel86) [inline webrob/neems/neem.py]: why not adding a flask handler for this error?

File: webrob/neems/neem.py [hunk 31]
```diff
@@ -18,27 +25,35 @@ def __init__(self,
...
+            self.acknowledgements = ''
+            self.environment = neem['environment']
+            self.activity = neem['activity']
+            self.agent = neem['agent']
+            self.keywords = neem['keywords']
+            self.image = neem['image']
+        except ConnectionFailure as e:
+            raise NEEMMetaException('An exception has occurred during connection with mongodb collection, please check!', exc=e)
+
+        except PyMongoError as e:
```
Comment (daniel86) [inline webrob/neems/neem.py]: why not adding a flask handler for this error?

File: webrob/neems/neem.py [hunk 32]
```diff
@@ -18,27 +25,35 @@ def __init__(self,
...
+            self.knowrob_tag = 'latest'
+            self.maintainer = neem['created_by']
+            self.authors = neem['created_by']
+            self.acknowledgements = ''
+            self.environment = neem['environment']
+            self.activity = neem['activity']
+            self.agent = neem['agent']
+            self.keywords = neem['keywords']
+            self.image = neem['image']
+        except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/neems/neem.py]: why not adding a flask handler for this error?

File: webrob/neems/manager.py [hunk 33]
```diff
@@ -41,5 +57,13 @@ def filter_neems(self, query_string):
...
-            {"$text": {"$search": query_string}}, {"_id": 1}))
+        try:
+            mongoDBMetaCollection = get_mongo_db_meta_collection()
+            return map(lambda doc: doc["_id"], mongoDBMetaCollection.find(
+                {"$text": {"$search": query_string}}, {"_id": 1}))
+
+        except ConnectionFailure as e:
+            raise NEEMMetaException('Failure connecting with mongodb with given credentials, please check inputs!', exc=e)
+
+        except PyMongoError as e:
```
Comment (daniel86) [inline webrob/neems/manager.py]: why not adding a flask handler for this error?

File: webrob/neems/manager.py [hunk 34]
```diff
@@ -41,5 +57,13 @@ def filter_neems(self, query_string):
...
         #   - regex search possible, but only without index!!
         #       db.meta.find({"keywords": {"$regex": ".*robot.*","$options": 'i'}},{_id: 1}).pretty()
-        return map(lambda doc: doc["_id"], mongoDBMetaCollection.find(
-            {"$text": {"$search": query_string}}, {"_id": 1}))
+        try:
+            mongoDBMetaCollection = get_mongo_db_meta_collection()
+            return map(lambda doc: doc["_id"], mongoDBMetaCollection.find(
+                {"$text": {"$search": query_string}}, {"_id": 1}))
+
+        except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/neems/manager.py]: why not adding a flask handler for this error?

File: webrob/neems/manager.py [hunk 35]
```diff
@@ -1,23 +1,39 @@
...
+                mongoDBMetaCollection = get_mongo_db_meta_collection()
+                self.neem_ids = mongoDBMetaCollection.find().distinct('_id')
+                # TODO: remove again multiply with 20 and return only self.neem_ids,
+                #  just for testing pagination as long as only few neems are there
+                self.neem_ids = self.neem_ids * 20
+
+            except ConnectionFailure as e:
+                raise NEEMMetaException('Failure connecting with mongodb with given credentials, please check inputs!', exc=e)
+
+            except PyMongoError as e:
```
Comment (daniel86) [inline webrob/neems/manager.py]: why not adding a flask handler for this error?

File: webrob/neems/manager.py [hunk 36]
```diff
@@ -1,23 +1,39 @@
...
+        count = get_settings_count()
+        if count > 0:
+            try:
+                mongoDBMetaCollection = get_mongo_db_meta_collection()
+                self.neem_ids = mongoDBMetaCollection.find().distinct('_id')
+                # TODO: remove again multiply with 20 and return only self.neem_ids,
+                #  just for testing pagination as long as only few neems are there
+                self.neem_ids = self.neem_ids * 20
+
+            except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/neems/manager.py]: why not adding a flask handler for this error?

File: webrob/pages/neems.py [hunk 37]
```diff
@@ -1,16 +1,30 @@
...
 @app.route('/neems')
 def render_neems():
+
+    # at first check if there is settings stored in db
+    try:
+        mongoDBMetaCollection = get_mongo_db_meta_collection()
+    except ConnectionFailure as e:
+        raise NEEMMetaException('An exception has occurred during connection with mongodb collection, please check!', exc=e)
+
+    except PyMongoError as e:
```
Comment (daniel86) [inline webrob/pages/neems.py]: why not adding a flask handler for this error?

File: webrob/pages/neems.py [hunk 38]
```diff
@@ -1,16 +1,30 @@
...
 
 neem_manager = NEEM_Manager()
 
 @app.route('/neems')
 def render_neems():
+
+    # at first check if there is settings stored in db
+    try:
+        mongoDBMetaCollection = get_mongo_db_meta_collection()
+    except ConnectionFailure as e:
```
Comment (daniel86) [inline webrob/pages/neems.py]: why not adding a flask handler for this error?

**Review by daniel86 — CHANGES_REQUESTED** [review]

File: webrob/pages/admin/db.py (hunk 25 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: here we do handle an error and points to flask handler.

File: webrob/pages/admin/db.py (hunk 26 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: NEEMMetaException is being raised and handled by a flask handler. or am I missing something here?

File: webrob/neems/manager.py (hunk 24 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py:19]: when application starts for the first time, it goes through manager and then subsequently in get_mongo_db_meta_collection but the exception thrown there results into an error. I am not 100% sure why this happens. Thus in order to avoid when db is either empty or not initialized, I would skip calling get_mongo_db_meta_collection.

File: webrob/pages/admin/db.py (hunk 25 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: done. Now ConnectionFailure has its own flask handler.

File: webrob/pages/admin/db.py (hunk 26 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: done. Now PyMongoError has its own flask handler.

File: webrob/pages/admin/db.py (hunk 28 above)
Comment (AbhijitVyas) [reply webrob/pages/admin/db.py]: done. Now SQLAlchemyError has its own flask handler.

File: webrob/neems/neem.py (hunk 30 above)
Comment (AbhijitVyas) [reply webrob/neems/neem.py]: done. Now SQLAlchemyError has its own flask handler.

File: webrob/neems/neem.py (hunk 31 above)
Comment (AbhijitVyas) [reply webrob/neems/neem.py]: done. Now PyMongoError has its own flask handler.

File: webrob/neems/neem.py (hunk 32 above)
Comment (AbhijitVyas) [reply webrob/neems/neem.py]: done. Now ConnectionFailure has its own flask handler.

File: webrob/neems/manager.py (hunk 34 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py]: done. Now ConnectionFailure has its own flask handler.

File: webrob/neems/manager.py (hunk 36 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py]: done. Now ConnectionFailure has its own flask handler.

File: webrob/pages/neems.py (hunk 38 above)
Comment (AbhijitVyas) [reply webrob/pages/neems.py]: done. Now ConnectionFailure has its own flask handler.

File: webrob/pages/neems.py (hunk 37 above)
Comment (AbhijitVyas) [reply webrob/pages/neems.py]: done. Now PyMongoError has its own flask handler.

File: webrob/neems/manager.py (hunk 33 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py]: done. Now PyMongoError has its own flask handler.

File: webrob/neems/manager.py (hunk 35 above)
Comment (AbhijitVyas) [reply webrob/neems/manager.py]: done. Now PyMongoError has its own flask handler.
